src/components/app.js

```diff
@@ -17,20 +21,56 @@ const {
   home, register, products, product, login,
 } = ROUTES;
 
-const App = () => (
-  <Router history={history}>
-    <div className="app-container">
-      <NavMenu />
-      <NavHeader history={history} />
-      <Switch>
-        <Route exact path={home} component={Home} />
-        <Route exact path={products} component={Products} />
-        <Route exact path={product} component={Product} />
-        <Route exact path={register} component={Register} />
-        <Route exact path={login} component={Login} />
-      </Switch>
-    </div>
-  </Router>
-);
-
-export default App;
+class App extends Component {
+  shouldComponentUpdate(nextProps) {
+    const { loginState } = this.props;
+    // deep comparisons
+    return !_.isEqual(loginState, nextProps.loginState);
+  }
+
+  render() {
+    const token = localStorage.getItem('OS_AUTH_TOKEN');
+
+    let isAuthenticated = false;
+    if (token) {
+      const decoded = jwtDecode(token);
+      if (Date.now() < decoded.exp * 1000) {
+        isAuthenticated = true;
+      } else {
+        localStorage.removeItem('OS_AUTH_TOKEN');
+      }
+    }
```

Comment:
The logic to check if the use is authenticated doesn't belong inside a component. This should be somewhere like an authentication helper to ensure separation of concerns and reusability.